src/core/reactor_backend.cc

```diff
@@ -1484,10 +1490,97 @@ class reactor_backend_uring final : public reactor_backend {
         delete pfd;
     }
     virtual future<std::tuple<pollable_fd, socket_address>> accept(pollable_fd_state& listenfd) override {
-        return _r.do_accept(listenfd);
+        if (listenfd.take_speculation(POLLIN | POLLOUT)) {
+            try {
+                listenfd.maybe_no_more_recv();
+                socket_address sa;
+                auto maybe_fd = listenfd.fd.try_accept(sa, SOCK_NONBLOCK | SOCK_CLOEXEC);
+                if (maybe_fd) {
+                    listenfd.speculate_epoll(EPOLLIN);
+                    pollable_fd pfd(std::move(*maybe_fd), pollable_fd::speculation(EPOLLOUT));
+                    return make_ready_future<std::tuple<pollable_fd, socket_address>>(std::move(pfd), std::move(sa));
+                }
+            } catch (...) {
+                return current_exception_as_future<std::tuple<pollable_fd, socket_address>>();
+            }
+        }
+        class accept_completion final : public io_completion {
+            pollable_fd_state& _listenfd;
+            socket_address _sa;
+            promise<std::tuple<pollable_fd, socket_address>> _result;
+        public:
+            accept_completion(pollable_fd_state& listenfd)
+                : _listenfd(listenfd) {}
+            void complete(size_t fd) noexcept final {
+                _listenfd.speculate_epoll(EPOLLIN);
+                pollable_fd pfd(file_desc::from_fd(fd), pollable_fd::speculation(EPOLLOUT));
+                _result.set_value(std::move(pfd), std::move(_sa));
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                auto set_ex_and_dispose = defer([eptr, this]() noexcept {
+                     _result.set_exception(eptr);
+                     delete this;
+                });
+                try {
+                    std::rethrow_exception(eptr);
+                } catch (const std::system_error& e) {
+                    if (e.code() == std::errc::invalid_argument) {
+                        try {
+                            // The chances are that we shutting down the connection.
+                            _listenfd.maybe_no_more_recv();
+                        } catch (...) {
+                            set_ex_and_dispose.cancel();
+                            _result.set_exception(std::current_exception());
+                            delete this;
+                        }
+                    }
+                } catch (...) {}
```

Comment:
It's less confusing IMO to move the contents of set_ex_and_dispose into the catch (...) block. Otherwise, the reader must wonder why we are ignoring the exception, then notice the defer thing.

Reply:
yeah, my eyes hurt when reading this function too. but we also need to replicate the body of `set_ex_and_dispose` in the else branch of `if (e.code() == std::errc::invalid_argument)`. i rewrote this function, hopefully it's less confusing now.